feat(vessels): clinking glasses

TimErmolt opened the pull request:
Можно чокаться сосудами для жидкостей с собутыльниками.
- Чтобы чокнуться, надо, держа сосуд в активной руке, поставить интент хелпа, прицелиться в кисть руки и кликнуть по другому мобу, у которого есть сосуд. У людей сосуд обязательно должен быть в руках, у киборгов -- активным модулем.
- Собутыльнику выводится радиальное меню с опцией принять предложение чокнуться или отказаться. Отсутствие выбора равно отказу.
![image](https://github.com/user-attachments/assets/8a94f047-1240-4054-86c9-e86ccca30187)
- Чокаться можно в принципе с любыми ливингами, если у них есть какой-либо сосуд. А почему нет?
- При успешном чокании содержимое двух сосудов немного перемешается (от 1 до 4 юнитов смеси жидкостей в сосудах). Этого не произойдёт, если хотя бы один из сосудов закрыт крышкой или хотя бы один пустой (иначе при определённых условиях можно дюпать жидкости).

- - Технически: добавил макрос `isvessel()` -- проверяет, является ли объект сосудом для жидкостей.
- - Технически: в [splash_text.dm](https://github.com/ChaoticOnyx/OnyxBay/blob/7584e1eac336af09096f350ef940be5b0e6e7323/code/modules/splash_text/splash_text.dm) при вызове функции показа сплеша не подавался аргумент на пропуск сообщения в чат. Добавил его, и в некоторых других местах оно тоже теперь нормально работает.

[Источник звука чоканья](https://freesound.org/people/elricadavis/sounds/764600/) опубликован под [лицензией CC0](https://creativecommons.org/publicdomain/zero/1.0/) (т.е. опубликовано как часть народного достояния). В метаданных файла я также указал автора и ссылку на источник.

close #5585 

<details>
<summary>Чейнджлог</summary>

```yml
🆑TheUnknownOne
rscadd: Теперь можно чокаться стаканами, бокалами, склянками и остальными сосудами для жидкостей. Держа сосуд в руке, прицельтесь в кисть руки (не важно, какую именно) и кликните по собутыльнику, тоже держащему сосуд. Собутыльнику нужно будет принять предложение в появившемся радиальном меню. Чокаться можно не только с людьми!
/🆑
```

</details>

- [x] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал(-а) все свои изменения и багов в них не нашёл(-ла).
- [x] Я запускал(-а) сервер со своими изменениями локально и все протестировал(-а).
- [x] Я ознакомился(-ась) c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

TimErmolt committed:
feat(vessels): clinking glasses

Issue #5585

Holding a vessel for liquids in hand, targeting a hand, and clicking on a mob that also has a vessel will allow you to clink glasses with them. Or with it. Yes, you can clink glasses with cyborgs and dionae, too. It's a feature.

SFX source: https://freesound.org/people/elricadavis/sounds/764600/ (licensed under CC0)

TimErmolt committed:
magic number to seconds

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
 
 	return FALSE
 
+/// Clinking glasses with your drinking buds. Also mixes up the vessels' contents a little. ~TimErmolt (TheUnknownOne)
```
Этот комментарий не имеет смысла, нормального названия достаточно.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+
+	var/obj/item/reagent_containers/vessel/target_vessel
+
+	// Try to prioritize the vessel in the hand we're targeting.
```
Это и так понятно, не требуется комментарий.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+
+	// Try to prioritize the vessel in the hand we're targeting.
+	if(ishuman(target))
+		var/mob/living/carbon/human/H = target
```
Перечитай, пожалуйста, вот этот [раздел](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md#%D0%BF%D1%80%D0%B0%D0%B2%D0%B8%D0%BB%D0%B0-%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D0%BF%D0%B5%D1%80%D0%B5%D0%BC%D0%B5%D0%BD%D0%BD%D1%8B%D1%85-%D0%B8-%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D0%B9).

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+	// Try to prioritize the vessel in the hand we're targeting.
+	if(ishuman(target))
+		var/mob/living/carbon/human/H = target
+		if(zone == BP_R_HAND && istype(H.r_hand, /obj/item/reagent_containers/vessel))
```
Добавь макрос `isvessel()`, будет полезен в будущем.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+		else if(istype(H.l_hand, /obj/item/reagent_containers/vessel))
+			target_vessel = H.l_hand
+
+	// If that doesn't work, just pick whatever we can find.
```
Я И ТАК ЭТО ПОНИМАЮ, НЕ НАДО КОММЕНТИРОВАТЬ КАЖДУЮ ПРОВЕРКУ. Кхм, прошу прощения.

Filatelele commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+	playsound(user.loc, 'sound/items/glasses_clink.ogg', 50, 1)
+
+	if(!is_open_container() || !target_vessel.is_open_container())
+		return TRUE // Can't mix contents if it ain't open
```
Я искренне не понимаю зачем комментировать очевидные вещи, но при этом не комментировать более неочевидные моменты. 

Ну ведь прекрасно прекрасно понятно что делает эта проверка исходя из названия!

Конечно ты можешь заниматься суходрочем, но вопрос зачем и нахуя все еще остается.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+	if(!target_vessel)
+		return FALSE
+
+	user.visible_message(SPAN_NOTICE("[user] reaches out to clink glasses with [target]..."),
```
Без сплешей не приму.

intercepti0n reviewed (changes requested):
Минус вайб.

Filatelele commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,71 @@
+	if(!target_vessel)
+		return FALSE
+
+	user.visible_message(SPAN_NOTICE("[user] reaches out to clink glasses with [target]..."),
```
В идеале нужно человеку выводить какой-нить ТГУИ алерт или прочее - чтобы он мог эксплицитно дать согласие или отказаться чокаться. Потому что этот спан нотис можно вообще пропустить, и пока ты обратишь внимание - две с половой секунды уже пройдут. 260 строчка с интентом чота выглядит жидковато для согласия, как мне кажется.

Roxreic replied:
бахнул радиальное меню на мобе который просит клинка и погнал

intercepti0n replied:
Норм идея, бтв.

TimErmolt committed:
remove some code comments

TimErmolt committed:
`isvessel()` macro

TimErmolt committed:
rename proc + radial menu offering

Rename the proc from `clink()` to `clink_glasses()` for clarity.
Change intent check when offering a clink to a radial menu selection for the target. Also adds radial menu icon states for the accept and decline buttons.
Add splash texts when offering or being offered a clink.

TimErmolt committed:
fix `force_skip_chat` in splash_text

The `force_skip_chat` argument was missing for some reason. Now added to the async invoke.

TimErmolt commented:
Окей, готово. Спасибо Сранклину за идею с радиальным меню.
![image](https://github.com/user-attachments/assets/8a94f047-1240-4054-86c9-e86ccca30187)
Заодно пофиксил, что сообщение в чат при показе сплеше не скипалось, когда должно было.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,72 @@
 	if(force && !(item_flags & ITEM_FLAG_NO_BLUDGEON) && user.a_intent == I_HURT)
 		return ..()
 
+	if(user.a_intent == I_HELP && (def_zone == BP_R_HAND || def_zone == BP_L_HAND) && clink_glasses(user, M, def_zone))
```
У тебя функция возаращает и так булево значение,  в чем проблема внутри почекать интент?

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,72 @@
+		user.visible_message(SPAN_NOTICE("[user] and [target] clink glasses!"),
+						 	 SPAN_NOTICE("You clink glasses with [target]! Cheers!"),
+						 	 SPAN_NOTICE("You hear glasses clinking."))
+		playsound(user.loc, 'sound/items/glasses_clink.ogg', 50, 1)
```
Задефайнь как SFX.

TimErmolt committed:
define SFX_GLASSES_CLINK

TimErmolt committed:
move checks inside proc

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,75 @@
+		else if(isvessel(human_target.l_hand))
+			target_vessel = human_target.l_hand
+
+	if(!target_vessel)
```
Ты не учитываешь случаи, когда чокаются с человеком и у него нет емкости.

TimErmolt replied:
Если нет в руках, оно будет искать на держателе костюма, на поясе, или где-нибудь ещё. Или в этом и есть проблема, нужно сделать, чтобы при чокании с человеком только на вещи в руках смотрело?

intercepti0n replied:
Если тебе так важно сделать возможность чокаться с **ДЖЕРЕЛО**, то делай `switch` и в нем чекай у человека **только** руки, у боргов актиный модуль, ну и у ливингов контетс, так уж и быть.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,75 @@
+	reagents.trans_to_holder(target_vessel.reagents, 1, transfer_amount, TRUE)
+	target_vessel.reagents.trans_to_holder(reagents, 1, transfer_amount, TRUE)
+
+	// It isn't exactly precise, but what do you expect when you're vigorously smashing two glasses together?
```
Все еще не вижу смысла в такого рода комментариях, no fun allowed.

TimErmolt replied:
В следующем коммите уберу.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,75 @@
+	user.visible_message(SPAN_NOTICE("[user] reaches out to clink glasses with [target]."),
+						 SPAN_NOTICE("You offer to clink glasses to [target]."))
+
+	user.show_splash_text(target, "offers to clink!", SPAN_NOTICE("[user] is offering to clink glasses."))
```
Очень странная компановка вызовов, я бы не ставил ньюлайна между двумя выводами сообщений в чат. Плюс не очень понимаю зачем делать вывод через `visible_message` и `show_splash_text`. Возможно, имеет смысл в начале этой всей цепочки выводить что-то по типу "поднимает бокал", а в конце общий визибл о том, что кто-то чекнулся, собственно. Не стоит разводить лишнюю графоманию, как мне кажется.

TimErmolt replied:
Тогда уберу ньюлайн, уберу сообщение в чат в `show_splash_text`

intercepti0n replied:
Не)

TimErmolt committed:
better mob type checks, remove unnecessary msgs

intercepti0n reviewed (changes requested):
Конфликты, плюс было бы круто сделать вывод сообщения о чоканьи через `show_splash_text_to_viewers`, чтобы было проще понять кто вообще и с кем произвел обмен жидкостями в стакане.

intercepti0n commented:
Плюс минусовой вайб за юзелезз коммент.

TimErmolt committed:
Merge branch 'dev' into clinking

TimErmolt committed:
change units to mls, add custom_emotes

- Replaces reagent units with milliliters (#13097)
- Replaces `visible_message()` procs with `custom_emote()`

TimErmolt commented:
В общем, обсудили в дискорде, сделал вместо `visible_message()` сообщения через `custom_emote()`. Ну и юниты жидкости заменил на миллилитры.

intercepti0n commented on code/modules/reagents/reagent_containers/vessel/_vessel.dm:
```diff
@@ -219,11 +219,80 @@
+	if(!reagents.total_volume || !target_vessel.reagents.total_volume)
+		return TRUE
+
+	var/transfer_amount = rand(5, 25) // milliliters
```
У нас так-то есть `units.dm` с дефайнами мер измерения.

TimErmolt replied:
Я вообще не заметил, что там MILLI есть. Ну, могу сделать `25 MILLI LITERS`, но такое ощущение, что не очень выглядит и легче просто комментарий оставить. Заодно не будет число умножаться лишний раз.

intercepti0n reviewed (approved):
Нейтральный вайб.

TobyThorne reviewed (approved):
доверюсь прошлому аппруверу если билд чокнется я не виноват
